packages/babel-traverse/src/scope/index.ts

```diff
@@ -838,24 +856,39 @@ export default class Scope {
         this.isPure(node.left, constantsOnly) &&
         this.isPure(node.right, constantsOnly)
       );
-    } else if (isArrayExpression(node)) {
+    } else if (isArrayExpression(node) || isTupleExpression(node)) {
       for (const elem of node.elements) {
-        if (!this.isPure(elem, constantsOnly)) return false;
+        if (elem !== null && !this.isPure(elem, constantsOnly)) return false;
       }
       return true;
-    } else if (isObjectExpression(node)) {
+    } else if (isObjectExpression(node) || isRecordExpression(node)) {
       for (const prop of node.properties) {
         if (!this.isPure(prop, constantsOnly)) return false;
       }
       return true;
+    } else if (isDecorator(node)) {
+      return this.isPure(node.expression, constantsOnly);
     } else if (isMethod(node)) {
       if (node.computed && !this.isPure(node.key, constantsOnly)) return false;
       if (node.kind === "get" || node.kind === "set") return false;
+      if (node.decorators) {
+        for (const decorator of node.decorators) {
+          if (!this.isPure(decorator, constantsOnly)) return false;
+        }
+      }
       return true;
-    } else if (isProperty(node)) {
+    } else if (isProperty(node) || isClassAccessorProperty(node)) {
       // @ts-expect-error todo(flow->ts): computed in not present on private properties
       if (node.computed && !this.isPure(node.key, constantsOnly)) return false;
-      return this.isPure(node.value, constantsOnly);
+      if (node.decorators) {
+        for (const decorator of node.decorators) {
+          if (!this.isPure(decorator, constantsOnly)) return false;
+        }
+      }
+      if (isObjectProperty(node) || node.static) {
```

Comment:
Evaluating a non-static field definition should be pure as long as it has no computed keys / fancy decorators, since the field initializer is not executed yet.